quickjs.c

```diff
@@ -6360,72 +6386,140 @@ static void build_backtrace(JSContext *ctx, JSValue error_obj,
                             int backtrace_flags)
 {
     JSStackFrame *sf;
-    JSValue str;
+    JSValue stack, prepare, saved_exception;
     DynBuf dbuf;
     const char *func_name_str;
     const char *str1;
     JSObject *p;
-    BOOL backtrace_barrier;
-    int flags;
+    JSFunctionBytecode *b;
+    BOOL backtrace_barrier, has_prepare;
+    JSRuntime *rt;
+    JSCallSiteData *csd;
+    int csd_count;
+    uint32_t i;
 
-    flags = JS_PROP_WRITABLE | JS_PROP_CONFIGURABLE;
-    js_dbuf_init(ctx, &dbuf);
-    if (filename) {
-        dbuf_printf(&dbuf, "    at %s", filename);
-        if (line_num != -1)
-            dbuf_printf(&dbuf, ":%d:%d", line_num, col_num);
-        dbuf_putc(&dbuf, '\n');
-        if (backtrace_flags & JS_BACKTRACE_FLAG_SINGLE_LEVEL)
+    rt = ctx->rt;
+    has_prepare = FALSE;
+    if (!rt->in_prepare_stack_trace && !JS_IsNull(ctx->error_ctor)) {
+        prepare = js_dup(ctx->error_prepare_stack);
+        has_prepare = JS_IsFunction(ctx, prepare);
+        rt->in_prepare_stack_trace = TRUE;
+    }
+
+    if (has_prepare) {
+        saved_exception = rt->current_exception;
+        rt->current_exception = JS_NULL;
+        i = 0;
+        csd_count = 8; // Sstart with a reasonable number.
```

Comment:
```suggestion
        csd_count = 8; // Start with a reasonable number.
```

Reply:
> Instead of js_realloc'ing a C array and building the CallSite array only at the end, construct and js_array_push() individual CallSite objects ASAP.

I tried that... kinda. I was using JS_DefineProperty, but even building the array will end up adding stack frames no? That's why I switched to doing it at the end.

Reply:
You mean when someone monkey-patches Array.prototype? JS_DefineProperty should be immune to that but yeah... js_array_push isn't because it calls JS_SetProperty, which invokes setters. Okay, disregard that suggestion then.